### ISSUE TITLE
gateway api status not reflected when running multiple istio version

### ISSUE THREAD

**Issue by kkk777-7** [istio/istio#52105]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

We use istio for gateway api.
We would like to employ canary versioning to validate istio behavior and run multiple versions of istio for a period of time.

we can deploy multiple istio envoys using the istio.io/rev label in Gateway resource, but only one revision of the istio is reflected in the Gateway resource status.
Therefore, it is currently not possible to have multiple versions of istio running in parallel for gateway api.

use istio version
```
$ istioctl tag list  
TAG    REVISION NAMESPACES
stable 1-20-8   
canary 1-22-2
```

gateway status
```
NAMESPACE       NAME              CLASS   ADDRESS         PROGRAMMED   AGE
istio-ingress   default-gateway   istio                   Unknown      49m
istio-ingress   specify-gateway   istio   10.195.47.182   True         49m

# default-gateway
Status:
  Conditions:
    Last Transition Time:  1970-01-01T00:00:00Z
    Message:               Waiting for controller
    Reason:                Pending
    Status:                Unknown
    Type:                  Accepted
    Last Transition Time:  1970-01-01T00:00:00Z
    Message:               Waiting for controller
    Reason:                Pending
    Status:                Unknown
    Type:                  Programmed

# specify-gateway
Status:
  Addresses:
    Type:   IPAddress
    Value:  10.195.47.182
  Conditions:
    Last Transition Time:  2024-07-17T04:50:04Z
    Message:               Resource accepted
    Observed Generation:   1
    Reason:                Accepted
    Status:                True
    Type:                  Accepted
    Last Transition Time:  2024-07-17T04:51:41Z
    Message:               Resource programmed, assigned to service(s) specify-gateway-istio.istio-ingress.svc.cluster.local:80
    Observed Generation:   1
    Reason:                Programmed
    Status:                True
    Type:                  Programmed
```
<details>

<summary>Gateway manifest</summary>

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: default-gateway
  namespace: istio-ingress
  labels:
    istio.io/rev: stable
spec:
  gatewayClassName: istio
  listeners:
  - name: default-gw
    hostname: "*.example.com"
    port: 80
    protocol: HTTP
    allowedRoutes:
      kinds:
      - group: gateway.networking.k8s.io
        kind: HTTPRoute
      namespaces:
        from: Selector
        selector:
          matchLabels:
            kubernetes.io/metadata.name: default-gw
```

```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: specify-gateway
  namespace: istio-ingress
  labels:
    istio.io/rev: canary
spec:
  gatewayClassName: istio
  listeners:
  - name: specify-gw
    hostname: "*.example.com"
    port: 80
    protocol: HTTP
    allowedRoutes:
      kinds:
      - group: gateway.networking.k8s.io
        kind: HTTPRoute
      namespaces:
        from: Selector
        selector:
          matchLabels:
            kubernetes.io/metadata.name: specify-gw
```
</details>

I followed the istiod logs and it seemed that only certain revisions of the pod were getting the `istio-gateway-status-leader` as the leader election.

version: can update gateway status
```
2024-07-17T05:00:54.504507Z     info    klog    successfully acquired lease istio-system/istio-gateway-status-leader
2024-07-17T05:00:54.504766Z     info    leader election lock obtained: istio-gateway-status-leader
2024-07-17T05:00:54.504779Z     info    Starting gateway status writer
2024-07-17T05:00:54.505320Z     info    klog    successfully acquired lease istio-system/istio-gateway-deployment-1-22-2
```

version: cannot update gateway status
```
2024-07-17T04:47:25.322462Z     info    klog    attempting to acquire leader lease istio-system/istio-gateway-status-leader...
2024-07-17T04:47:25.322510Z     info    klog    attempting to acquire leader lease istio-system/istio-gateway-deployment-1-20-8...
```

### Version

```Text
$ istioctl version
client version: 1.22.2
istiod version: 1.20.8
istiod version: 1.22.2
data plane version: 1.20.8 (1 proxies), 1.22.2 (1 proxies)
$ kubectl version     
Client Version: v1.30.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.4
```

### Additional Information

[bug-report.tar.gz](https://github.com/user-attachments/files/16260650/bug-report.tar.gz)

**Comment by kkk777-7** [comment]
maybe I think...

gateway deployment controller use `NewPerRevisionLeaderElection` ,  so can deploy every revision without any problem.
https://github.com/istio/istio/blob/master/pilot/pkg/bootstrap/configcontroller.go#L169
https://github.com/istio/istio/blob/master/pilot/pkg/leaderelection/leaderelection.go#L225-L228

but, gateway status writer use `NewLeaderElection`, so can not reflect gateway status.
https://github.com/istio/istio/blob/master/pilot/pkg/bootstrap/configcontroller.go#L149

Is there any reason why NewPerRevisionLeaderElection cannot be used in gateway status writer?

**Comment by kkk777-7** [comment]
@howardjohn 
Hi, the following PR shows that the deployment controller uses NewPerRevisionLeaderElection.
https://github.com/istio/istio/pull/44746

Is there any reason why not using it in status writer?
If you know of any, please let me know when you have time.

**Comment by howardjohn** [comment]
This is the same root cause as https://github.com/istio/istio/issues/52438  fwiw

**Comment by howardjohn** [comment]
I believe this should be fixed by https://github.com/istio/istio/pull/55715

**Comment by hiwayama** [comment]
Hello. 
I faced the issue which it occurs too many requests to k8s API when I'm using k8s Gateway API and running multiple istiod.

Is this related this issue?

## Version
```
canary: 1.27.0
stable: 1.27.1
```

I used k8s Gateway API such as Gateway/GRPCRoute/HTTPRoute resources.

The requests are for `PUT GRPCRoute/HTTPRoute`.
I guess that both istiod try to update the same resource's status.

Let me know if this is the known issue or I should create this as a new issue.
